src/lib/storage/constraints/single_constraint_checker.hpp

```diff
@@ -0,0 +1,70 @@
+#pragma once
+
+#include <algorithm>
+#include <optional>
+#include <string>
+#include <vector>
+
+#include "operators/validate.hpp"
+#include "storage/constraints/row_templated_constraint_checker.hpp"
+#include "storage/segment_accessor.hpp"
+
+namespace opossum {
+
+template <typename T>
+class SingleConstraintChecker : public RowTemplatedConstraintChecker<T> {
+ public:
+  SingleConstraintChecker(const Table& table, const TableConstraintDefinition& constraint)
+      : RowTemplatedConstraintChecker<T>(table, constraint) {
+    Assert(constraint.columns.size() == 1, "Only one column constraints allowed for SingleConstraintChecker");
+  }
+
+  virtual std::shared_ptr<std::vector<T>> get_inserted_rows(std::shared_ptr<const Table> table_to_insert) const {
+    auto values = std::make_shared<std::vector<T>>();
+
+    for (const auto& chunk : table_to_insert->chunks()) {
+      const auto segment = chunk->segments()[this->_constraint.columns[0]];
+      const auto chunk_segment_accessor = create_segment_accessor<T>(segment);
+      for (ChunkOffset chunk_offset = 0; chunk_offset < chunk->size(); chunk_offset++) {
+        std::optional<T> value = chunk_segment_accessor->access(chunk_offset);
+        if (value.has_value()) {
+          values->emplace_back(value.value());
+        }
+      }
+    }
+    return values;
+  }
+
+  virtual bool preprocess_chunk(std::shared_ptr<const Chunk> chunk) {
+    const auto segment = chunk->segments()[this->_constraint.columns[0]];
+    this->segment_accessor = create_segment_accessor<T>(segment);
```

Comment:
Just saw this now: Using the state of the object is a weird way to pass information between two methods. Why not return the accessor here and return nullptr if none is needed?

Reply:
I understand that it seems wrong somehow. Unfortunately returning the segment accessor here to the base class won't be better, as the segment accessor is needed as state only by the subclasses single constraint checker. The variant for multiple unique columns uses different information as state here, e.g. a vector of segments.

Reply:
This sounds like a good reason to split up the method as suggested in the other comment. Compactness of code is not necessarily the number one priority.

Reply:
Done.

Reply:
Ok, just splitting it up did not fix the issue though. The variables in the objects are really a huge OOP smell. :/

The only thing I can come up with right now is returning a context object (upcasted to a generic type) from the `preprocess_chunk` method that is later downcasted in the specific implementations.

Reply:
I agree on the OOP smell, but returning a context object that is passed back to the specific implementations (and especially casted again for how many times?) doesn't seem like a nice solution to me as well. What about passing back a function object from `prepare_read_chunk` that is called for each offset of the chunk to get the rows? We could put the required segment accessors etc. into that object. And if the chunk shouldn't be checked, we can just return a `nullptr` or empty optional.

Reply:
`std::function` is horribly slow. Calling it once per row is probably inefficient.

Remember that downcasts are free if you use static_cast.

If you want to stick with the current implementation, it should be documented better. First, I'd add `_cache` to those ugly member variables and rename the method that sets them to include that, too. Second, the members and the calls should get a comment explaining what we are doing and why we can't come up with anything better.

Reply:
Not sure if you missed this remaining comment.

Reply:
Documented that behavior a bit more and renamed the member variables. Adding that "cache" meaning to the name of the methods felt a bit weird and I'm not sure if the names like that are really good, but is probably a good idea to emphasize that behavior. The methods using the cached state are now called:

- `prepare_read_chunk_cached`
- `is_cached_chunk_check_required`
- `get_row_from_cached_chunk`

I feel these names are un-handy. Maybe you have some suggestions?